test/recipes/70-test_cert_cb.t

```diff
@@ -0,0 +1,36 @@
+#! /usr/bin/env perl
+# Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+#
+# Licensed under the Apache License 2.0 (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+use strict;
+use OpenSSL::Test qw/:DEFAULT cmdstr srctop_file/;
+use OpenSSL::Test::Utils;
+use TLSProxy::Proxy;
+
+my $test_name = "test_cert_cb";
+setup($test_name);
+
+plan skip_all => "TLSProxy isn't usable on $^O"
+    if $^O =~ /^(VMS)$/;
+
+plan skip_all => "Test only supported in a shared build" if disabled("shared");
+
+my $proxy = TLSProxy::Proxy->new(
```

Comment:
TLSProxy doesn't seem the best choice for this kind of test. TLSProxy is mainly useful where we want to modify or examine what is actually happening "on the wire". Here we are testing that a particular API is working, i.e. that the certificate callback gets called when we expect it to. This seems more like an sslapi test to me...and in fact we already have a test to do this: "test_cert_cb". So, I'm not sure what this test gives us over what "test_cert_cb" already does.

What would be more useful would be to modify test_cert_cb to confirm that the sig algs are available.

Reply:
The s_client/s_server tests use SSL_check_chain() in the cert_cb, whereas the current one doesn't. This shows up as a simple used-to-work-now-it-doesn't test in 1.1.1 vs 1.1.1a as described in issue #9099.

Reply:
In which case we are doing a test to confirm that SSL_check_chain() works inside a certificate callback. That's still an API test and better suited to sslapitest rather than TLSProxy. It works for TLSProxy, but only by an accident of the way s_server/s_client is implemented. This is a brittle test. If someone modifies the way s_server/s_client works then the test no longer works and we wouldn't necessarily know it. We are much better off testing this explicitly by modifying/extending test_cert_cb (or adding a new test if appropriate).

Reply:
@mattcaswell not disagreeing with you. Our initial internal test added SSL_chain_check() to `test_cert_cb`, but because this was demonstrable via s_client/s_server, @kaduk decided to go the TLSProxy route.

Reply:
I think this is the wrong approach. TLSProxy uses s_server/s_client as a convenience but I'd like to change it to use something else at some point. The purpose of TLSProxy is to enable us to modify or inspect bytes on the wire. That isn't the case here.

Reply:
I have an sslapitest that will reproduce the s_server behavior here (not quite polished enough to commit yet, though) -- it took me a while to realize that the  key property we need is to have no certificate at all on the server SSL if SSL_check_chain fails, which in turn causes tls_choose_sigalg()  to fail to find anything usable and bail on the connection.